### DISCUSSION TITLE
Why doesn't getMatchInfo take a TokenSource rather than a CharSequence?

### DISCUSSION THREAD

**Discussion by vsajip** [congo-cc/congo-parser-generator#81] [Q&A] [answered]
The signature
https://github.com/congo-cc/congo-parser-generator/blob/feb0e8bf2b6ba9b3970b855874a2283c173feefb/src/templates/java/Lexer.java.ftl#L207
suggests it can be called with any `CharSequence input`, but in practice it is only ever called with a lexer instance:
https://github.com/congo-cc/congo-parser-generator/blob/feb0e8bf2b6ba9b3970b855874a2283c173feefb/src/templates/java/Lexer.java.ftl#L292-L299
Wouldn't it be better to just use `TokenSource input` instead? What do we really gain by making `TokenSource` implement `CharSequence`? Asking because I'm looking at implementing the polyglot versions - in C#, for example, there isn't a `CharSequence` -equivalent interface, and I'm dubious of adding an interface analogous to Java's just for this.

**Comment by revusky** [comment, marked as answer]
Well, the basic idea behind the overall refactoring is that we can optionally use a `CharSequence` (basically a `String`, but it could be `StringBuilder` or whatever) in that method, not just a `TokenSource`. And, as you can see, the `getMatchInfo` method is static, so it doesn't rely on any state that exists in the `TokenSource` (basically the lexer). So, in principal, we can call the `getMatchInfo` method just statically without even having any instance of the Lexer (or Parser) around. 

Generally, I was toying with the idea of getting rid of any *state* in the XXXLexer class. But there is still `lexicalState` and `activeTokenTypes`. Those are the only instance variables in the XXXLexer template. (By default, anyway. The application programmer can always inject some extra ones.) My actual intention in this whole refactoring was to get `lexicalState` and `activeTokenTypes` into the XXXParser object, and then just pass them into the lexer. And that way, the XXXLexer would be  completely state-free.

As for `position`, which previously was a kind of bookmark or cursor that maintained the information of where we were at in the lexing, but that is actually gone now. The information of where to start scanning for the next token is taken as being in the last scanned token itself. Basically, the key method to understand would be: https://github.com/congo-cc/congo-parser-generator/blob/main/src/templates/java/Lexer.java.ftl#L156-L176

In particular, see line 170 there, which shows you that the tokenizing machinery just picks up at the `endOffset` of the preceding token which is passed in. One rather elegant aspect of this is that, for example, if you truncate the token, for example, via: `tok.truncate(1)` in a token hook, let's say, which is BTW just syntactic sugar for `tok.setEndOffset(tok.getEndOffset()-1)` then the next call to `getNextToken()` will pick up tokenization at the *adjusted* end of the last token. Previously, when you did stuff like this, you had to make sure to adjust the `position` variable in the XXXLexer object. But now there is no `position` variable instance variable. The *position* at which to start scanning the next token is in the token itself!

Now, it is true that the `TokenSource` superclass does contain state, but most of that is stuff that is set up at initialization, all the stuff of where the various lines start/end and so on. It does also maintain a sort of token cache and location table. which is state, I grant. Actually, I was thinking about making `TokenSource` a separate class, but somehow (for reasons I don't honestly recall) I decided to make it an abstract base class for any XXXLexer. Though, actually, come to think of it, originally FreeCC had a separate object called `FileLineMap` and the XXXLexer object had an instance of that and it delegated to `FileLineMap` to get information about line numbers and so on. But then I decided to make that just part of XXXLexer and then later separated it out into an abstract base class. And, as you know, the tokens do not now store any line/column info, but just beginning and ending offset. And then when line/column information is needed for human consumption, it just delegates to the XXXLexer object from which it was vended, which is effectively the `TokenSource`, previously known as `FileLineMap`. 

Well, that is the basis of all the massive refactoring. I think that it would be good if we could get rid of the remaining instance variables in XXXLexer, namely `activeTokenTypes` and `lexicalState`. So if you want to experiment with that, by all means.

Now, as for `TokenSource` implementing the `CharSequence` interface, most of the value of that is because core objects, particularly `java.lang.String` implement that interface, so there is the possibility (though it takes some careful design) of being able to just pass in a String or any *stringy* sort of object in that spot and seeing what the next token would be at any position. I mean in a static context... And I think that's quite potentially useful. But, as for what to do in Python or CSharp, well, I'm not sure. I guess I would have to think about it....

Anyway, I hope the above comments are useful.

By the way, have you looked at the latest version of lexer.py.ftl ? You may notice something novel about it! (Though it has really nothing to do with this topic of conversation!)

Oh, and another thing that almost slipped my mind. The whole thing with beginTokenOffset as opposed to position... I think that has to do with supporting the `MORE` construct. If you have a `MORE`, i.e pending characters from a previous match then...
